File: 39. Combination Sum.cpp
class Solution {
public:
    void combs(int idx,vector<int>&arr,int target,int n,vector<int>l,vector<vector<int>>&lst){
        if(idx>n){
            if(target==0) lst.push_back(l);
            return;
        }
        l.push_back(arr[idx]);
        
        if(arr[idx]<=target){
            combs(idx,arr,target-arr[idx],n,l,lst);
        }

        l.pop_back();
        combs(idx+1,arr,target,n,l,lst);
        
    }
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        int n=candidates.size()-1;
        vector<vector<int>>final;
        vector<int>lst;
        combs(0,candidates,target,n,lst,final);
        return final;
    }
};

File: 39. Combination Sum.py
class Solution:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        def combsums(idx,lst,n,arr,target,final):
            if idx>n:
                if target==0:
                    final.append(lst[:])
                return

            lst.append(arr[idx])

            if arr[idx]<=target:
                
                combsums(idx,lst,n,arr,target-arr[idx],final)
            lst.pop()
            combsums(idx+1,lst,n,arr,target,final)
            
        lst=[]
        n=len(candidates)-1
        final=[]
        combsums(0,lst,n,candidates,target,final)
        return final
